src/baldr/graphtileheader.cc

```diff
@@ -12,16 +13,23 @@ namespace valhalla {
 namespace baldr {
 
 // Default constructor.
-GraphTileHeader::GraphTileHeader() {
-  memset(this, 0, sizeof(GraphTileHeader));
-  strncpy(version_, PACKAGE_VERSION, kMaxVersionSize);
-  version_[kMaxVersionSize - 1] = 0;
-  tile_size_ = 0;
+GraphTileHeader::GraphTileHeader()
+    : // initialization of bitfields done here in c++20 can be done in the class definition
+      graphid_(0), density_(0), name_quality_(0), speed_quality_(0), exit_quality_(0),
+      has_elevation_(0), has_ext_directededge_(0), nodecount_(0), directededgecount_(0),
+      predictedspeeds_count_(0), spare1_(0), transitioncount_(0), spare3_(0), turnlane_count_(0),
+      spare4_(0), transfercount_(0), spare2_(0), departurecount_(0), stopcount_(0), spare5_(0),
+      routecount_(0), schedulecount_(0), signcount_(0), spare6_(0), access_restriction_count_(0),
+      admincount_(0), spare7_(0) {
+  set_version(PACKAGE_VERSION);
 }
 
 // Set the version string.
 void GraphTileHeader::set_version(const std::string& version) {
-  strncpy(version_, version.c_str(), kMaxVersionSize);
+  // reinitializing the version array before copying
+  version_ = {};
+  std::copy(version.begin(), version.begin() + std::min(kMaxVersionSize, sizeof(version.c_str())),
```

Comment:
the previous code was a bug because someone could screw with the version string and make it shorter than `kMaxVersionSize` and `strncpy` would have read past the end, so its good you are fixing it however i think you have a new bug here, this should be:

```suggestion
  std::copy(version.begin(), version.begin() + std::min(kMaxVersionSize, version.size()),
```

otherwise it will only ever copy 8 bytes because `c_str()` is a pointer which is 64bits wide (these days)

Reply:
nice catch :-)